### ISSUE TITLE
getShadowRoot Browser hangups

### ISSUE THREAD

**Issue by jasonparallel** [bitwarden/clients#10151]
### Steps To Reproduce

The removal of the || node.childNodes.length !== 0) check in getShadowRoot made here https://github.com/bitwarden/clients/commit/bbf14730222408292a2e00dd15c7b4ccc74cba8d#diff-8ff1ef69519807ae86fc872374b223656c9459d1b8b0304e8a2b602182b68f98
is causing 1 sec+ browser hangups when it runs on a select box with a large number of select options.

Also the comment on the method is not longer accurate. "Will return null if the node is not an HTMLElement or if the node has child nodes."

### Expected Result

Don't hang up the browser

### Actual Result

The browser hangs up

### Operating System

macOS

### Web Browser

Chrome

### Build Version

6.1 (did not happen in 5.2)

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by Neonwarden** [comment]
Hi there,

I am unable to reproduce this issue, it has been escalated for further investigation. If you have more information that can help us, please add it below. 

Thanks!

**Comment by jasonparallel** [comment]
I'll see if I can create a small example page.

**Comment by jasonparallel** [comment]
I modified the bitwarden plugin (pulled from github) to add timing around chrome.dom.openOrClosedShadowRoot(node)
open  https://codepen.io/jasonparallel/pen/oNrbWod
with bitwarden thinking it has a password for codepen.io (not sure if this is needed)
I got a 4174 ms lock up on openOrClosedShadowRoot
<img width="763" alt="Screenshot 2024-07-17 at 3 46 33 PM" src="https://github.com/user-attachments/assets/0e2b84ee-6869-41ff-b53a-d537d1c46cfc">

Prior to the referenced change, nodes like this would not have openOrClosedShadowRoot run on them.

**Comment by jasonparallel** [comment]
@Neonwarden Was that helpful?
@cagonzalezcs Did you change intent to no longer exclude nodes with child nodes?

**Comment by cagonzalezcs** [comment]
@jasonparallel 

I need to update the comment on that method.

So the reversal on that is indeed purposefully done, and the behavior you're describing was resolved with the `deepQuery` methodology introduced with https://github.com/bitwarden/clients/pull/9063

If you look at the thread mentions at the bottom of the PR, however, we ran into some heavy issues on websites that deeply nest ShadowDOM elements. As a result, we've hardcoded a feature flag reversal of the reworked implementation. 

The reason why we're once again checking for child nodes is because of examples such as the one referenced in this Github issue - https://github.com/bitwarden/clients/issues/9566

The form reference in that thread has a custom web component that contains both a shadow root and a regular node. In effect, discarding nodes that contain a child breaks that form's autofill behavior. 

I'll be revisiting the performance improvements soon. When a PR for that comes into play, I'll try to remember to post a link to it here.

**Comment by ThatBeaverDev** [comment]
I'm experiencing something that may be related to this with https://turbowarp.org. Image attached below (taking 300ms for the function call)
![Screenshot_2024-10-19_at_12 26 43](https://github.com/user-attachments/assets/7ac91d77-dc67-404c-b910-2003e258aeb1)

**Comment by TTTPOB** [comment]
Found this issue disturbing when using a very reasonable website, ncbi blast, an academic tool for biology, https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome.
trying to open this page cause the whole page freeze for a long time (at least 30 seconds, in below devtools screenshot, I clicked stop before I can operate the page to save some time, so it is less than it).

![Image](https://github.com/user-attachments/assets/ed44746d-02da-4ddd-85d8-a49ca1d7ffa4)

**Comment by tszming** [comment]
@TTTPOB Yes, your example shows exactly what the problem is — it's very slow when a page contains very complicated DOM. I also use 1Password in the same browser, and it doesn't have this issue.

<img width="731" alt="Image" src="https://github.com/user-attachments/assets/5cb94314-6071-4e07-aa4c-4b9108a619f7" />

**Comment by aciccarello** [comment]
I'm also seeing this issue on an internal company site. 25 seconds of the main thread being locked. I can't share the site but hopefully these performance tracking details are helpful.

<img width="1182" height="687" alt="Image" src="https://github.com/user-attachments/assets/dca8c4a9-afa9-49cf-8043-9937dd43014b" />

<img width="772" height="426" alt="Image" src="https://github.com/user-attachments/assets/3bafaab1-47bd-4853-bad5-1e45e52235c8" />

Extension version: 2025.6.1
Browser: Chrome 140.0.7259.2 (Official Build) dev (arm64)

**Comment by Staindk** [comment]
This exact same thing has just started affecting at least myself and one colleague on an internal company website. The problem page has a bunch of dropdowns with 1000s of options in each dropdown. I'm on Chrome, think my colleague is as well, and we both use the Bitwarden extension.

Dev tools performance tab looks the same as the commenter above.

No issue when disabling the extension.

**Comment by chelming** [comment]
Seeing this on servicenow instances. 25 second thread time. 

<img width="772" height="205" alt="Image" src="https://github.com/user-attachments/assets/5a66a43c-6407-4158-a3d7-5dc8ff61d108" />

<img width="493" height="78" alt="Image" src="https://github.com/user-attachments/assets/ebb73fdf-d8b7-40de-a80d-eb0864ea3fdc" />
